programs/_updown/_updown.8
'\" t
.\"     Title: _UPDOWN
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 10/06/2010
.\"    Manual: [FIXME: manual]
.\"    Source: [FIXME: source]
.\"  Language: English
.\"
.TH "_UPDOWN" "8" "10/06/2010" "[FIXME: source]" "[FIXME: manual]"
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ipsec__updown \- kernel and routing manipulation script
.SH "SYNOPSIS"
.PP
\fI_updown\fR
is invoked by pluto when it has brought up a new connection\&. This script is used to insert the appropriate routing entries for IPsec operation on some kernel IPsec stacks, such as KLIPS and MAST, and may do other necessary work that is kernel or user specific, such as defining custom firewall rules\&. The interface to the script is documented in the pluto man page\&.
.SH "VARIABLES"
.PP
The
\fI_updown\fR
is passed along a number of variables which can be used to act differently based on the information:
.PP
\fBPLUTO_VERSION\fR
.RS 4
indicates what version of this interface is being used\&. This document describes version 1\&.1\&. This is upwardly compatible with version 1\&.0\&.
.RE
.PP
\fBPLUTO_VERB\fR
.RS 4
specifies the name of the operation to be performed, which can be one of
\fBprepare\-host\fR,
\fBprepare\-client\fR,
\fBup\-host\fR,
\fBup\-client\fR,
\fBdown\-host\fR
or
\fBdown\-client\fR\&. If the address family for security gateway to security gateway communications is IPv6, then a suffix of \-v6 is added to this verb\&.
.RE
.PP
\fBPLUTO_CONNECTION\fR
.RS 4
is the name of the connection for which we are routing\&.
.RE
.PP
\fBPLUTO_NEXT_HOP\fR
.RS 4
is the next hop to which packets bound for the peer must be sent\&.
.RE
.PP
\fBPLUTO_INTERFACE\fR
.RS 4
is the name of the ipsec interface to be used\&.
.RE
.PP
\fBPLUTO_ME\fR
.RS 4
is the IP address of our host\&.
.RE
.PP
\fBPLUTO_MY_CLIENT\fR
.RS 4
is the IP address / count of our client subnet\&. If the client is just the host, this will be the host\'s own IP address / max (where max is 32 for IPv4 and 128 for IPv6)\&.
.RE
.PP
\fBPLUTO_MY_CLIENT_NET\fR
.RS 4
is the IP address of our client net\&. If the client is just the host, this will be the host\'s own IP address\&.
.RE
.PP
\fBPLUTO_MY_CLIENT_MASK\fR
.RS 4
is the mask for our client net\&. If the client is just the host, this will be 255\&.255\&.255\&.255\&.
.RE
.PP
\fBPLUTO_PEER\fR
.RS 4
is the IP address of our peer\&.
.RE
.PP
\fBPLUTO_PEER_CLIENT\fR
.RS 4
is the IP address / count of the peer\'s client subnet\&. If the client is just the peer, this will be the peer\'s own IP address / max (where max is 32 for IPv4 and 128 for IPv6)\&.
.RE
.PP
\fBPLUTO_PEER_CLIENT_NET\fR
.RS 4
is the IP address of the peer\'s client net\&. If the client is just the peer, this will be the peer\'s own IP address\&.
.RE
.PP
\fBPLUTO_PEER_CLIENT_MASK\fR
.RS 4
is the mask for the peer\'s client net\&. If the client is just the peer, this will be 255\&.255\&.255\&.255\&.
.RE
.PP
\fBPLUTO_MY_PROTOCOL\fR
.RS 4
lists the protocols allowed over this IPsec SA\&.
.RE
.PP
\fBPLUTO_PEER_PROTOCOL\fR
.RS 4
lists the protocols the peer allows over this IPsec SA\&.
.RE
.PP
\fBPLUTO_MY_PORT\fR
.RS 4
lists the ports allowed over this IPsec SA\&.
.RE
.PP
\fBPLUTO_PEER_PORT\fR
.RS 4
lists the ports the peer allows over this IPsec SA\&.
.RE
.PP
\fBPLUTO_MY_ID\fR
.RS 4
lists our id\&.
.RE
.PP
\fBPLUTO_PEER_ID\fR
.RS 4
lists our peer\'s id\&.
.RE
.PP
\fBPLUTO_PEER_CA\fR
.RS 4
lists the peer\'s CA\&.
.RE
.SH "SEE ALSO"
.PP
\fBipsec\fR(8),
\fBipsec_pluto\fR(8)\&.
.SH "HISTORY"
.PP
Man page written for the Linux FreeS/WAN project <\m[blue]\fBhttp://www\&.freeswan\&.org/\fR\m[]> by Michael Richardson\&. Original program written by Henry Spencer\&.


CHANGES

Note that listings are in chronological order of release times, not in order
of version numbers, so you will find 2.5.x and 2.6.x releases intersperced.

See also docs/KNOWN_BUGS and http://bugs.openswan.org/

v2.6.30 (unreleased)
* Add geode-aes to the list of crypto modules to load [Paul]
* Add --listen/listen= option to limit listening to a single IP [Paul]
* sa_policy_bit_names was missing ModeConfig DNS and WINS bit names [Paul]
* SAREF: Add sareftrack=<no|yes|conntrack> connection option [Paul]
* Fix for: ASSERTION FAILED at connections.c:1579: c->kind == CK_TEMPLATE [Paul]
* Fix for "handling event EVENT_RETRANSMIT for <invalid>" [Paul]
* Fix for specifying protport=47 (GRE has no ports) [Paul]
* Bugtracker bugs fixed:
   # 252/619 more than 20 payloads in message; ignored [paul]
   # 690 ipsec lwdnsq --help not implemented
   # 860 Port --random for newhostkey [Paul]
   #1005 Incorrect message "R_U_THERE_ACK has unexpected sequence number" [Mike]
   #1040: Fix to compile without DEBUG [Paul]
   #1054: Startup warning: "ignored obsolete keyword (null)" [Michael Smith]
   #1112: Prototypes only, if function enabled in c-source with KLIPS or
          PFKEY [Henry N.]
   #1115: Fix various warnings u_char * vs. char * for sscanf,... [Henry N.]
   #1149: pluto uses empty NAT_OA as IDci, Server behind NAT and non natted
          Windows XP [Wolfgang Nothdurft]
   #1151: The ipsec module is not removed by 'ipsec setup stop' [Paul]

v2.6.29 (September 27, 2010)
! This release is made for CVE-2010-3302 and CVE-2010-3308
! Duplicate CVE's issues for these are CVE-2010-3752 and CVE-2010-3753
* XAUTH: Avoid buffer overflow in CISCO DNS info [dhr/paul]
         Avoid shell problems with single quotes CISCO DNS paramters [dhr/paul]
  http://www.cve.mitre.org/cgi-bin/cvename.cgi?name=2010-3302
* XAUTH: Avoid buffer overflow in CISCO BANNER [dhr/paul]
         Avoid shell problems with single quotes in CISCO paramters [dhr/paul]
  http://www.cve.mitre.org/cgi-bin/cvename.cgi?name=2010-3308
* NETKEY: Fix for spurious %hold netlink-acquires [Paul/dhr]
* KLIPS: Fix compiling on 2.6.18 based RHEL5 kernels [Paul]
* Various fixes based on automated source code review [dhr]
* SAREF: Updated for 2.6.35 kernel [Harald]
* KLIPS: Updated for 2.6.35 kernel [Harald]
* PACKAGING Use Epoch 1: for Debian/Ubuntu [Simon]
* MAST: fix iptables rule "leak" on rekey [Bart]
* MAST: use only the most recent iptables rule [Bart]
* pluto: restrict rekeymargin to be smaller than salifetime [Bart]
* MAST: ensure we don't end up with mtu=0 on mast0 [Bart/Paul]
* MAST: enforce outgoing tunnel policy [Bart]
* MAST: use addflow pfkey command to set policy on tunnel SAs [Bart]
* Added a new pfkey flag, POLICYONLY, to the ADDFLOW command [Bart]
* MAST: allow for setting of policy for inbound SAs [Bart]
* MAST: favour deleting an SA even if the pfkey op failed [Bart]
* HAVESTATSD: Log new phase2 messages as a result of a rekey [Paul]
* MAST: use iptables --comment to show the conn name [Bart]
* VNET: differentiate instantiation of road warriors and vnet [Paul]
* Log LEAK_DETECTIVE and HAVE_LIBNSS support on startup [Paul]
* [IKEv2] connections were broken since 2.6.25 [Avesh]
* MAST: new "ipsec policy" command replaces "ipsec eroute" [Bart]
* Fix SElinux warning in realsetup (bz628879) [Avesh]
* Support for SHA2_256 in IKEv2 (bz621790) [Avesh]
* IKEv2: Fix for using MD5 and PRF conversion function [Avesh]
* SAREF: Improved workaround for rp_filter [Bart]
* NSS: Increase minimum nss for rhbz#453577 [Paul]
  (this allows us to revert workaround in git 6c8ff2791d1)
* SAREF: Added /proc/net/ipsec/saref that shows kernel patch state [Bart]
* SAREF: Add SAref checks to ipsec verify [bart/paul]
* A bunch of Makefile.inc variables could not be overridden properly [Paul]
* Bugtracker bugs fixed:
   #1042 NF_INET_LOCAL_OUT vs NF_IP_LOCAL_OUT still broken [David]
   #1102: Ensure KLIPS still works on older kernels [Paul]
   #1113: crypt_dh.c: Fix "warning: multi-line comment" [Tuomo]

v2.6.28 (July 29, 2010)
* MAST: Fix SAref vmalloc() call that could cause a kernel panic [Bart]
* SAREF: rework exported interface to avoid might_sleep() during rcu lock [Bart]
* SAREF: Use the correct mask when updating nfmark in rcv path [Bart]
* MAST: Cleanup updown.mast iptables rule management [Bart]
* MAST: Rework mast init scripts to use conntrack [Bart]
* MAST: Remove iptables rules after SA is deleted [Bart]
* SAREF: Fix bug in stream-socket saref mode [Bart]
* SAREF: Ported to apply on Linux 2.6.34 [Harald]
* MAST: Fix for NAT-T mode (set ixs->outgoing_said in mast mode) [Bart]
* MAST: Fix module unload with mast [Bart]
* Allow rightsubnet=vnet:%priv rightprotoport=17/%any w/o right=%any [Paul]
* SAREF: Log SAref and SAbind capabilities on pluto startup [Paul]
* Log tunnel down with HAVE_STATSD as "down", not "unknown" [Paul]
* Changes to _realsetup.in for making the init script LSB compliant [Avesh]
  rhbz #594767
* _startnetkey update for DNS and NetworkManager [Avesh]
* SAREF: fix bug in stream-socket saref mode [Bart]
* Added TCP header flags to ipsec_print_ip() output [Bart]
* KLIPS: Don't fail non-existant header_ops (breaks ppp) [David]
* KLIPS: ip_select_ident hashing fix in AH xmit path [Kirill Berezin]
* HAVE_STATSD: Fix phase1/phase2 logging through HAVE_STATSD interface [David]
* HAVE_STATSD: log output fix when two connections share a phase1 [David]
* HAVE_STATSD: Slightly clarified and changed log messages [Paul]
* SAREF: Clarified defines and fixed nfmark printing in HAVE_STATSD [dhr/Paul]
* MAST: Temp workaround in _updown.mast for martians problem [Bart/Paul]
* Cleanup source code to use C-style comments [Tuomo]
* Bugtracker bugs fixed:
   # 1120: [PATCH] netlink receive buffer size too small for linux 2.6.32
           [Roman Hoog Antink]

v2.6.27 (June 21, 2010)
* Fixes to the SAREF / MAST code to avoid recursion loops [Bart]
* KLIPS compile fixes for 2.4 kernels [David]

contrib/_updown/_updown.c
/* $Id: _updown.c,v 1.1 2005/07/21 20:42:57 ken Exp $
 *
 * THIS IS OBSOLETE - It is no longer functionally identical to the scripts
 *
 *
 * This program is replacement for /usr/local/lib/ipsec/_updown script
 * and its functionality is identical.
 *
 * Installation:
 * 1. Compile with "gcc -O2 -o updown updown.c"
 * 2. Install "cp -f updown /usr/local/lib/ipsec"
 * 3. Update your configs so that they include leftupdown and rightupdown
 * keywords, e.g.:
 *
 * conn test
 *   ...
 *   leftupdown=/usr/local/lib/ipsec/updown
 *   rightupdown=/usr/local/lib/ipsec/updown
 *   ...
 * 
 * Characteristics:
 * - written in C, thus faster and less resource intensive than shell script
 * - uses iptables
 * - doesn't yet support opportunistic encryption
 *
 * Written by Pawel Krawczyk <kravietz at aba.krakow.pl>
 *
 * License: GPLv2.
 */

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <errno.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

char *load(const char *what);
int my_system(char *bin, char **argv);

int main(void) {
	char *pluto_verb;
	char *pluto_peer_client;
	char *pluto_interface;
	char *pluto_me;
	char *pluto_my_client;
	char *argv[20];
	int status;
	int testing = 0;

	if( load("UPDOWN_TESTING") == NULL) {
		chdir("/etc");
	} else
		testing = 1;

	/* Wczytujemy zmienne przekazane nam przez Pluto */
	pluto_verb = load("PLUTO_VERB");
	if(pluto_verb == NULL) { 
		fprintf(stderr, "PLUTO_VERB not set\n");
		return 1;
	}
	pluto_peer_client = load("PLUTO_PEER_CLIENT");
	if(pluto_peer_client == NULL) {
		fprintf(stderr, "PLUTO_PEER_CLIENT not set\n");
		return 1;
	}
	pluto_interface = load("PLUTO_INTERFACE");
	if(pluto_interface == NULL) {
		fprintf(stderr, "PLUTO_INTERFACE not set\n");
		return 1;
	}
	pluto_me = load("PLUTO_ME");
	if(pluto_me == NULL) {
		fprintf(stderr, "PLUTO_ME not set\n");
		return 1;
	}
	pluto_my_client = load("PLUTO_MY_CLIENT");
	if(pluto_my_client == NULL) {
		fprintf(stderr, "PLUTO_MY_CLIENT not set\n");
		return 1;
	}

	/* Dodajemy lub usuwamy routing w zaleznosci od
	 * polecenia przekazanego w PLUTO_VERB
	 */
	if(strncmp(pluto_verb, "route-", 6) == 0 ||
	   strncmp(pluto_verb, "up-", 3) == 0) {

		argv[0]="/bin/ip"; argv[1]="route"; argv[2]="add";
		argv[3]=pluto_peer_client;
		argv[4]="dev"; argv[5]=pluto_interface;
		argv[6]="via"; argv[7]=pluto_me;
		argv[8]=0;
		if(!testing) {
			status = my_system("/bin/ip", argv);
			if(status != 0) return status;
		}
		else
			printf("route add %s\n", pluto_peer_client);
	}

	if(strncmp(pluto_verb, "unroute-", 8) == 0 ||
	   strncmp(pluto_verb, "down-", 5) == 0) {

		argv[0]="/bin/ip"; argv[1]="route"; argv[2]="del";
		argv[3]=pluto_peer_client; argv[4]="dev";
		argv[5]=pluto_interface; argv[6]="via"; argv[7]=pluto_me;
		argv[8]=0;
		if(!testing) {
			status = my_system("/bin/ip", argv);
			if(status != 0) return status;
		}
		else
			printf("route del %s\n", pluto_peer_client);

	}
	
	if(strncmp(pluto_verb, "prepare-", 8) == 0) {

		argv[0]="/bin/ip"; argv[1]="route"; argv[2]="del";
		argv[3]=pluto_peer_client; argv[4]=0;
		if(!testing) {
			/* We ignore any errors from this command,
			 * as it's used to clear up any routes that
			 * may or may not be present.
			 */
			int null = open("/dev/null", O_WRONLY);
			dup2(null, 1); dup2(null, 2);
			status = my_system("/bin/ip", argv);
		}
		else
			printf("prepare del %s\n", pluto_my_client);

	}

	return 0;
}

char *load(const char *what) {
	char *tmp, *tmp1;

	tmp1 = getenv(what);
	if(tmp1 == NULL)
		return NULL;
	
	tmp = strchr(tmp1, '=');
	if(tmp != NULL) {
		tmp++;
		tmp1 = tmp;
	}

	return tmp1;
}

int my_system(char *bin, char **argv) {
	int status, pid;

	pid = vfork();
	if(pid == -1)
		return 1;
	if(pid == 0) {
		execv(bin, argv);
		return 127;
	}
	do {
		if(waitpid(pid, &status, 0) == -1) {
			if(errno != EINTR)
				return 1;
		} else
			return status;
	} while(1);
}


